File: 卡码/代码随想录/C++/链表/06-链表相交/LeetCode_02_07.cpp
//面试题 02.07. 链表相交
struct ListNode{
	int val;
	ListNode* next;
	ListNode():val(0),next(nullptr){}
	ListNode(int val):val(val),next(nullptr){}
	ListNode(int val,ListNode* next):val(val),next(next){}
};

class Solution {
public:
  ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
  	ListNode *tmpA=headA,*tmpB=headB;
  	int lenA=0,lenB=0,gap=0;
  	ListNode *maxNode,*minNode;
  	while(tmpA){
			lenA++;
			tmpA=tmpA->next;
		}
		while(tmpB){
			lenB++;
			tmpB=tmpB->next;
		}
		if(lenA>lenB){
			gap=lenA-lenB;
			maxNode=headA;
			minNode=headB;
		}else{
			gap=lenB-lenA;
			maxNode=headB;
			minNode=headA;
		}
		while(gap--) maxNode=maxNode->next;
		while(maxNode){
			if(maxNode==minNode) return maxNode;
			maxNode=maxNode->next;
			minNode=minNode->next;
		}
		return nullptr;
  }
};

File: 卡码/代码随想录/Python/链表/06-链表相交/LeetCode_02_07.py
#面试题 02.07. 链表相交
from typing import *

class ListNode:
	def __init__(self,val=0,next=None):
		self.val=val
		self.next=next

class Solution:
	def getIntersectionNode(self,headA:ListNode,headB:ListNode)->ListNode:
		lenA,lenB,gap=0,0,0
		tmpA,tmpB=headA,headB
		while tmpA!=None:
			lenA+=1
			tmpA=tmpA.next
		while tmpB!=None:
			lenB+=1
			tmpB=tmpB.next
		if lenA>lenB:
			gap=lenA-lenB
			maxNode=headA
			minNode=headB
		else:
			gap=lenB-lenA
			maxNode=headB
			minNode=headA
		while gap>0:
			gap-=1
			maxNode=maxNode.next
		while maxNode!=None:
			if maxNode==minNode:
				return maxNode
			maxNode=maxNode.next
			minNode=minNode.next
		return None
